src/V3Timing.cpp

```diff
@@ -0,0 +1,766 @@
+                                                                       refp->cloneTree(false)}),
+                            SenItem);
+            });
+        }
+        // Create an always with the assignment under it (assign the RHS to the intermediate)
+        flp->warnOff(V3ErrorCode::BLKSEQ, true);
+        auto* const alwaysp
+            = new AstAlways{flp, VAlwaysKwd::ALWAYS, new AstSenTree{flp, senItemsp},
+                            new AstAssign{flp, new AstVarRef{flp, newvscp, VAccess::WRITE},
+                                          nodep->rhsp()->unlinkFrBack()}};
+        alwaysp->user1SetOnce();  // Do not revisit this always, so it doesn't get transformed into
+                                  // an initial
+        controlp->replaceWith(alwaysp);
+        alwaysp->addStmtp(controlp);
+        // Assign the intermediate to the LHS after the timing control
+        auto* const assignp = new AstAssign{flp, nodep->lhsp()->unlinkFrBack(),
+                                            new AstVarRef{flp, newvscp, VAccess::READ}};
+        assignp->user1SetOnce();  // So we don't do the net delay again
+        nodep->replaceWith(assignp);
+        VL_DO_DANGLING(nodep->deleteTree(), nodep);
+    }
+    virtual void visit(AstWait* nodep) override {
+        FileLine* const flp = nodep->fileline();
+        AstNode* senItemsp = nullptr;
+        // Wait on changed events related to the vars in the wait statement
+        {
+            const VNUser4InUse user4InUse;
+            nodep->condp()->foreach<AstVarRef>([&](AstVarRef* refp) {
+                AstVarScope* const vscp = refp->varScopep();
+                if (vscp->user4SetOnce()) return;
+                const bool isEvent = vscp->dtypep() && vscp->dtypep()->basicp()
+                                     && vscp->dtypep()->basicp()->isEvent();
+                const auto edgeType = isEvent ? VEdgeType::ET_EVENT : VEdgeType::ET_CHANGED;
+                senItemsp = AstNode::addNext(
+                    senItemsp,
+                    new AstSenItem{flp, edgeType, new AstVarRef{flp, vscp, VAccess::READ}});
+            });
+        }
+        AstNode* const condp = nodep->condp()->unlinkFrBack();
+        AstNode* const bodysp
+            = nodep->bodysp() ? nodep->bodysp()->unlinkFrBackWithNext() : nullptr;
+        AstNode* newStmtsp;
+        if (senItemsp) {
+            // Put the event control in a while loop with the wait expression as condition
+            newStmtsp = new AstWhile{
+                flp, new AstLogNot{flp, condp},
+                new AstEventControl{flp, new AstSenTree{flp, VN_AS(senItemsp, SenItem)}, nullptr}};
+        } else {
+            // We have to await forever instead of simply returning in case we're deep in a
+            // callstack
+            auto* const awaitp = new AstCAwait{flp, new AstCStmt{flp, "VlForever{}"}};
+            awaitp->statement(true);
+            newStmtsp = new AstIf{flp, new AstLogNot{flp, condp}, awaitp};
```

Comment:
Shouldn't we be able to always statically evaluate the condition here? (And maybe warn that the condition is constant.) If not, I'm curious why.

Reply:
We should be able to, but `V3Const` will do it anyway. Is there some nice function in Verilator that determines if a constant expression is true?

Reply:
`V3Const::constifyEdit(condp)` should return the result node, which should be `AstConst` at this point.